src/tests/performance/connections.rs

```diff
@@ -0,0 +1,216 @@
+    // ├─────────────┼─────────┼──────────────┼──────────────┼──────────────┼──────────────┼──────────────┼────────────┤
+    // │          50 │      20 │           19 │            1 │            0 │            0 │            0 │       3.86 │
+    // ├─────────────┼─────────┼──────────────┼──────────────┼──────────────┼──────────────┼──────────────┼────────────┤
+    // │          50 │      30 │           21 │            9 │           14 │            0 │            0 │       5.81 │
+    // ├─────────────┼─────────┼──────────────┼──────────────┼──────────────┼──────────────┼──────────────┼────────────┤
+    // │          50 │      50 │           20 │           30 │           19 │            0 │            0 │       9.53 │
+    // └─────────────┴─────────┴──────────────┴──────────────┴──────────────┴──────────────┴──────────────┴────────────┘
+
+    // maximum time allowed for a single iteration of the test
+    const MAX_ITER_TIME: Duration = Duration::from_secs(20);
+
+    /// maximum peers to configure node with
+    const MAX_PEERS: u16 = 50;
+
+    let synth_counts = vec![1u16, 5, 10, 20, 30, 50];
+
+    let mut all_stats = Vec::new();
+
+    let target = TempDir::new().expect("couldn't create a temporary directory");
+    // start node
+    let mut node = Node::builder()
+        .max_peers(MAX_PEERS as usize)
+        .start(target.path(), NodeType::Stateless)
+        .await
+        .unwrap();
+    let node_addr = node.addr();
+
+    for synth_count in synth_counts {
+        // setup metrics recorder
+        let test_metrics = TestMetrics::default();
+        // register metrics
+        metrics::register_counter!(METRIC_ACCEPTED);
+        metrics::register_counter!(METRIC_TERMINATED);
+        metrics::register_counter!(METRIC_REJECTED);
+        metrics::register_counter!(METRIC_ERROR);
+
+        let mut synth_handles = Vec::with_capacity(synth_count as usize);
+        let mut synth_exits = Vec::with_capacity(synth_count as usize);
+        let (handshake_tx, mut handshake_rx) =
+            tokio::sync::mpsc::channel::<()>(synth_count as usize);
+
+        let test_start = Instant::now();
+
+        // start synthetic nodes
+        for _ in 0..synth_count {
+            let (exit_tx, exit_rx) = tokio::sync::oneshot::channel::<()>();
+            synth_exits.push(exit_tx);
+
+            let synth_handshaken = handshake_tx.clone();
+            // Synthetic node runs until it completes or is instructed to exit
+            synth_handles.push(tokio::spawn(async move {
+                tokio::select! {
+                    _ = exit_rx => {},
+                    _ = simulate_peer(node_addr, synth_handshaken) => {},
+                };
+            }));
+        }
+
+        // Wait for all peers to indicate that they've completed the handshake portion
+        // or the iteration timeout is exceeded.
+        let _ = tokio::time::timeout(MAX_ITER_TIME, async move {

```

Comment:
From the comment it sounds as it could be implemented with tokio's Barrier?

Reply:
The idea here is to wait until all synth_nodes complete handshakes or timeout occur but without blocking the threads. Each thread when complete handshake is receiving messages sent by the rippled just to ensure we won't suffer any overflows due to incoming messages. Barrier would block also block thread till all other threads complete their handshakes.
So to achieve this there is tokio::mpsc where synth_node sends empty value and main thread just wait to receive anything from each thread.